cache/redis_cache.go

```diff
@@ -106,76 +98,222 @@ func (r *redisCache) Get(key *Key) (*CachedData, error) {
-	if err != nil {
-		return 0, nil
+	b := make([]byte, 0, n+4)
+	b = append(b, byte(n>>24), byte(n>>16), byte(n>>8), byte(n))
+	b = append(b, s...)
+	return b
+}
+
+func (r *redisCache) stringFromBytes(bytes []byte) (string, int) {
+	b := bytes[:4]
+	n := uint32(b[3]) | (uint32(b[2]) << 8) | (uint32(b[1]) << 16) | (uint32(b[0]) << 24)
+	s := bytes[4 : 4+n]
+	return string(s), int(4 + n)
+}
+
+func (r *redisCache) metadataToByte(contentMetadata *ContentMetadata) []byte {
+	cLength := contentMetadata.Length
+	cType := r.stringToBytes(contentMetadata.Type)
+	cEncoding := r.stringToBytes(contentMetadata.Encoding)
+	b := make([]byte, 0, len(cEncoding)+len(cType)+8)
+	b = append(b, byte(cLength>>56), byte(cLength>>48), byte(cLength>>40), byte(cLength>>32), byte(cLength>>24), byte(cLength>>16), byte(cLength>>8), byte(cLength))
+	b = append(b, cType...)
+	b = append(b, cEncoding...)
+	return b
+}
+
+func (r *redisCache) metadataFromByte(b []byte) (*ContentMetadata, int) {
+	cLength := uint64(b[7]) | (uint64(b[6]) << 8) | (uint64(b[5]) << 16) | (uint64(b[4]) << 24) | uint64(b[3])<<32 | (uint64(b[2]) << 40) | (uint64(b[1]) << 48) | (uint64(b[0]) << 56)
+	offset := 8
+	cType, sizeCType := r.stringFromBytes(b[offset:])
+	offset += sizeCType
+	cEncoding, sizeCEncoding := r.stringFromBytes(b[offset:])
+	offset += sizeCEncoding
+	metadata := &ContentMetadata{
+		Length:   int64(cLength),
+		Type:     cType,
+		Encoding: cEncoding,
 	}
+	return metadata, offset
+}
 
+func (r *redisCache) Put(reader io.Reader, contentMetadata ContentMetadata, key *Key) (time.Duration, error) {
+	medatadata := r.metadataToByte(&contentMetadata)
 	ctx, cancelFunc := context.WithTimeout(context.Background(), putTimeout)
 	defer cancelFunc()
-	err = r.client.Set(ctx, key.String(), marshalled, r.expire).Err()
-
+	stringKey := key.String()
+	err := r.client.Set(ctx, stringKey, medatadata, r.expire).Err()
 	if err != nil {
 		return 0, err
 	}
-
+	// we don't read all the reader content then send it in one call to redis to avoid memory issue
+	// if the content is big (which is the case when chproxy users are fetching a lot of data)
+	buffer := make([]byte, 2*1024*1024)
+	for {
+		n, err := reader.Read(buffer)
+		// the reader should return an err = io.EOF once it has nothing to read or at the last read call with content.
+		// But this is not the case with this reader so we check the condition n == 0 to exit the read loop.
+		// We kept the err == io.EOF in the loop in case the behavior of the reader changes
+
+		if n == 0 {
+			break
+		}
+		if err != nil && !errors.Is(err, io.EOF) {
+			return 0, err
+		}
+		err = r.client.Append(ctx, stringKey, string(buffer[:n])).Err()
+		if err != nil {
+			return 0, err
+		}
+		if errors.Is(err, io.EOF) {
+			break
+		}
+	}
 	return r.expire, nil
 }
 
 func (r *redisCache) Name() string {
 	return r.name
 }
 
-func toBytes(stream io.Reader) ([]byte, error) {
-	buf := new(bytes.Buffer)
+type RedisStreamReader struct {
```

Comment:
it can be private I think
```suggestion
type redisStreamReader struct {
```